fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1281,10 +1283,11 @@ def _init_param_attributes(self, p: Parameter) -> None:
             # shard in pinned memory so that we can do a non-blocking transfer.
             # This is only needed during training and not evaluation.
             if self.ssd_offload:
-                assert isinstance(p, SsdFlatParameter)
+                assert isinstance(p, ssd_offload.SsdParameter)
                 # Gradients also need to be offloaded to SSD otherwise it can result in
                 # OOMs when the memory requirements of a model are larger than host memory.
                 p._cpu_grad = ssd_offload.SsdTensorHandle.from_tensor(torch.zeros_like(p.data, device="cpu"))
+                p._cpu_grad.allow_unsafe_changes = True
```

Comment:
Is this set to True so that we can check properties of the tensor?

Reply:
allow_unsafe_changes permits FSDP change the metadata of a tensor in some (at least to my mind) unsafe ways. It can:
    * Alter the tensor shape (the tensor is originally created to be the full one, and then .data is overridden to store only the local shard
    * Change .data even if the storage state is ON_CPU_DIRTY (FSDP _shard_parameters_)
    * Change requires_grad (when FSDP replaces .data with cuda equivalent, requires_grad is set to false)